{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### RNA reads processing (Illumina 2x125bp),  TPM (DNA, RNA) calculation, gene of interest annotation, and taxonomy identification"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1.Concatenate and rename files\n",
    "source /nesi/project/uoa02469/Software/PipelineSetup/genomic.sh\n",
    "\n",
    "mkdir comb-fastq\n",
    "OUT=comb-fastq\n",
    "\n",
    "for r in R1 R2;\n",
    "do\n",
    "    cat *4462-37*_${r}_001.fastq.gz > $OUT/RNA-S1-Sed1_${r}.fastq.gz\n",
    "    cat *4462-38*_${r}_001.fastq.gz > $OUT/RNA-S1-Sed2_${r}.fastq.gz\n",
    "    cat *4462-39*_${r}_001.fastq.gz > $OUT/RNA-S1-Sed3_${r}.fastq.gz\n",
    "    cat *4462-41*_${r}_001.fastq.gz > $OUT/RNA-S2-Sed1_${r}.fastq.gz\n",
    "    cat *4462-42*_${r}_001.fastq.gz > $OUT/RNA-S2-Sed2_${r}.fastq.gz\n",
    "    cat *4462-43*_${r}_001.fastq.gz > $OUT/RNA-S2-Sed3_${r}.fastq.gz\n",
    "    cat *4462-45*_${r}_001.fastq.gz > $OUT/RNA-S3-Sed1_${r}.fastq.gz\n",
    "    cat *4462-46*_${r}_001.fastq.gz > $OUT/RNA-S3-Sed2_${r}.fastq.gz\n",
    "    cat *4462-47*_${r}_001.fastq.gz > $OUT/RNA-S3-Sed3_${r}.fastq.gz\n",
    "    cat *4462-49*_${r}_001.fastq.gz > $OUT/RNA-S4-Sed1_${r}.fastq.gz\n",
    "    cat *4462-50*_${r}_001.fastq.gz > $OUT/RNA-S4-Sed2_${r}.fastq.gz\n",
    "    cat *4462-51*_${r}_001.fastq.gz > $OUT/RNA-S4-Sed3_${r}.fastq.gz\n",
    "    cat *4462-53*_${r}_001.fastq.gz > $OUT/RNA-S5-Sed1_${r}.fastq.gz\n",
    "    cat *4462-54*_${r}_001.fastq.gz > $OUT/RNA-S5-Sed2_${r}.fastq.gz\n",
    "    cat *4462-55*_${r}_001.fastq.gz > $OUT/RNA-S5-Sed3_${r}.fastq.gz\n",
    "    cat *4462-57*_${r}_001.fastq.gz > $OUT/RNA-S6-Sed1_${r}.fastq.gz\n",
    "    cat *4462-58*_${r}_001.fastq.gz > $OUT/RNA-S6-Sed2_${r}.fastq.gz\n",
    "    cat *4462-59*_${r}_001.fastq.gz > $OUT/RNA-S6-Sed3_${r}.fastq.gz\n",
    "    cat *4462-61*_${r}_001.fastq.gz > $OUT/RNA-S7-Sed1_${r}.fastq.gz\n",
    "    cat *4462-62*_${r}_001.fastq.gz > $OUT/RNA-S7-Sed2_${r}.fastq.gz\n",
    "    cat *4462-63*_${r}_001.fastq.gz > $OUT/RNA-S7-Sed3_${r}.fastq.gz\n",
    "    cat *4462-40*_${r}_001.fastq.gz > $OUT/RNA-Filt1_${r}.fastq.gz\n",
    "    cat *4462-44*_${r}_001.fastq.gz > $OUT/RNA-Filt2_${r}.fastq.gz\n",
    "    cat *4462-48*_${r}_001.fastq.gz > $OUT/RNA-Filt3_${r}.fastq.gz\n",
    "    cat *4462-52*_${r}_001.fastq.gz > $OUT/RNA-Filt4_${r}.fastq.gz    \n",
    "    cat *4462-56*_${r}_001.fastq.gz > $OUT/RNA-Filt5_${r}.fastq.gz\n",
    "    cat *4462-60*_${r}_001.fastq.gz > $OUT/RNA-Filt6_${r}.fastq.gz\n",
    "    cat *4462-64*_${r}_001.fastq.gz > $OUT/RNA-Filt7_${r}.fastq.gz\n",
    "    cat *4462-65*_${r}_001.fastq.gz > $OUT/RNA-Filt8_${r}.fastq.gz  \n",
    "    cat *4462-66*_${r}_001.fastq.gz > $OUT/RNA-Filt9_${r}.fastq.gz   \n",
    "done\n",
    "\n",
    "mkdir fastqc/\n",
    "mkdir fastqc/raw\n",
    "\n",
    "module load FastQC/0.11.7\n",
    "for i in RNA*fastq;do fastqc $i -o fastqc/raw -f fastq -t 4; done\n",
    "\n",
    "cd fastqc/raw\n",
    "module load MultiQC/1.7-gimkl-2018b-Python-3.7.3\n",
    "multiqc ./ --interactive\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 2. sortmerna choose only non-rRNA\n",
    "\n",
    "mkdir 2.sortmerna\n",
    "\n",
    "#sortmerna slurm script\n",
    "#job=17566460 and 17566472\n",
    "\n",
    "export FILES=($(ls -1 OG4462B_fastq/comb-fastq/1.comb/RNA-*_R1.fastq.gz)) && NUM=${#FILES[@]} && ZNUM=$(($NUM - 1))\n",
    "if [ $ZNUM -ge 0 ]; then sbatch --array=0-$ZNUM sortmerna1.sbatch; fi\n",
    "#!/bin/bash\n",
    "#SBATCH -A ga02676\n",
    "#SBATCH -J 1.sortrna\n",
    "#SBATCH --time 8:00:00\n",
    "#SBATCH --mem-per-cpu=2000\n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=20\n",
    "#SBATCH --requeue\n",
    "#SBATCH -e slurm-sortrna.%A_%a.err\n",
    "#SBATCH -o slurm-sortrna.%A_%a.out\n",
    "\n",
    "export FILES=($(ls -1 OG4462B_fastq/comb-fastq/1.comb/RNA-*_R1.fastq.gz))\n",
    "FILENAME=${FILES[$SLURM_ARRAY_TASK_ID]}\n",
    "SAMPLE=$(basename ${FILENAME} _R1.fastq.gz);\n",
    "export DIR=OG4462B_fastq/comb-fastq/1.comb\n",
    "    \n",
    "module load BBMap/38.81-gimkl-2020a\n",
    "    \n",
    "reformat.sh in1=$DIR/${SAMPLE}_R1.fastq.gz in2=$DIR/${SAMPLE}_R2.fastq.gz out=$DIR/${SAMPLE}_in.fastq.gz;\n",
    "\n",
    "gunzip $DIR/${SAMPLE}_in.fastq.gz\n",
    "  \n",
    "module load SortMeRNA/2.1-gimkl-2017a\n",
    "sortmerna --ref sortmerna-2.1/rRNA_databases/silva-bac-16s-id90.fasta,sortmerna-2.1/index/silva-bac-16s-db:\\\n",
    "sortmerna-2.1/rRNA_databases/silva-bac-23s-id98.fasta,sortmerna-2.1/index/silva-bac-23s-db:\\\n",
    "sortmerna-2.1/rRNA_databases/silva-arc-16s-id95.fasta,sortmerna-2.1/index/silva-arc-16s-db:\\\n",
    "sortmerna-2.1/rRNA_databases/silva-arc-23s-id98.fasta,sortmerna-2.1/index/silva-arc-23s-db:\\\n",
    "sortmerna-2.1/rRNA_databases/silva-euk-18s-id95.fasta,sortmerna-2.1/index/silva-euk-18s-db:\\\n",
    "sortmerna-2.1/rRNA_databases/silva-euk-28s-id98.fasta,sortmerna-2.1/index/silva-euk-28s:\\\n",
    "sortmerna-2.1/rRNA_databases/rfam-5s-database-id98.fasta,sortmerna-2.1/index/rfam-5s-db:\\\n",
    "sortmerna-2.1/rRNA_databases/rfam-5.8s-database-id98.fasta,sortmerna-2.1/index/rfam-5.8s-db \\\n",
    "--reads $DIR/${SAMPLE}_in.fastq --num_alignments 1 --fastx --paired_in --aligned sortmerna/${SAMPLE}_rRNA --other sortmerna/mrna.${SAMPLE}_non_rRNA --log -v -a 20\n",
    "\n",
    "#repair.sh in=sortmerna/mrna.${SAMPLE}_non_rRNA.fastq \\\n",
    "    out1=sortmerna/mrna.${SAMPLE}_R1.fastq out2=sortmerna/mrna.${SAMPLE}_R2.fastq \\\n",
    "    outs=sortmerna/mrna.${SAMPLE}_single.fastq\n",
    "    \n",
    "#remove empty line in sortmerna output, issue refers to https://github.com/biocore/sortmerna/issues/117\n",
    "for i in sortmerna/mrna.RNA-*; \n",
    "    do export sample=$(echo ${i} | sed -e 's/sortmerna\\/\\(.*\\)/\\1/g'); \n",
    "    sed -i '/^$/d' $i ;\n",
    "    done\n",
    "\n",
    "mv sortmerna OG4462B_fastq/comb-fastq/2.sortmerna\n",
    "\n",
    "cd OG4462B_fastq/comb-fastq/\n",
    "    \n",
    "for i in 2.sortmerna/*.log;do echo $i;grep 'Total' $i;done\n",
    "\n",
    "module load FastQC/0.11.7\n",
    "mkdir fastqc/mrna\n",
    "    \n",
    "for i in 2.sortmerna/mrna.*fastq;do fastqc $i -o fastqc/mrna -f fastq -t 4; done\n",
    "    \n",
    "module load MultiQC/1.7-gimkl-2018b-Python-3.7.3\n",
    "cd fastqc\n",
    "multiqc fastqc/mrna/ --interactive\n",
    "cd ../"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 3. Quality filter and trim\n",
    "\n",
    "mkdir 3.trimmomatic\n",
    "echo \">FastQC_adapter\" > iua.fna\n",
    "echo \"AGATCGGAAGAG\" >> iua.fna\n",
    "\n",
    "#testing with iua.fna with 5:20\n",
    "for i in 2.sortmerna/mrna.RNA*_R1.fastq; do  name=$(basename ${i} _R1.fastq);     \n",
    "    trimmomatic PE -threads 12 -phred33 -quiet \\\n",
    "    2.sortmerna/${name}_R1.fastq 2.sortmerna/${name}_R2.fastq \\\n",
    "    3.trimmomatic/t30.q.${name}_R1.fastq.gz \\\n",
    "    3.trimmomatic/t30.q.${name}_R1.single1.fastq.gz \\\n",
    "    3.trimmomatic/t30.q.${name}_R2.fastq.gz \\\n",
    "    3.trimmomatic/t30.q.${name}_R2.single2.fastq.gz \\\n",
    "    ILLUMINACLIP:iua.fna:1:25:7   SLIDINGWINDOW:4:30 MINLEN:60; \n",
    "done\n",
    "\n",
    "#ran fastqc again \n",
    "mkdir fastqc/trim_mrna\n",
    "\n",
    "module load FastQC/0.11.7\n",
    "for i in 3.trimmomatic/t30.*fastq.gz;do fastqc $i -o fastqc/trim_mrna -f fastq -t 4; done\n",
    "cd fastqc/\n",
    "\n",
    "module load MultiQC/1.7-gimkl-2018b-Python-3.7.3\n",
    "multiqc trim_mrna/ --interactive\n",
    "\n",
    "#but still some remaining bad bases at the end, crop last 10 bases\n",
    "\n",
    "cd ../3.trimmomatic/\n",
    "\n",
    "for filename in t30.q.*_R1.fastq.gz;\n",
    "do\n",
    " i=$(basename ${filename} _R1.fastq.gz);\n",
    "\n",
    "    trimmomatic PE -threads 6 -phred33 -quiet \\\n",
    "        ${i}_R1.fastq.gz ${i}_R2.fastq.gz \\\n",
    "        new.${i}_R1.fastq new.${i}_R1.single1.fastq \\\n",
    "        new.${i}_R2.fastq new.${i}_R2.single2.fastq \\\n",
    "        CROP:115 SLIDINGWINDOW:4:30 MINLEN:60\n",
    "\n",
    "done\n",
    "\n",
    "#final fastqc\n",
    "mkdir fastqc/final-trim_mrna\n",
    "\n",
    "module load FastQC/0.11.7\n",
    "for i in 3.trimmomatic/new.*fastq;do fastqc $i -o fastqc/final-trim_mrna -f fastq -t 4; done\n",
    "cd fastqc/\n",
    "\n",
    "module load MultiQC/1.7-gimkl-2018b-Python-3.7.3\n",
    "multiqc final-trim_mrna --interactive"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 3. deduplication slurm script\n",
    "export FILES=($(ls -1 /home/htee663/ga02676/Waiwera_project/2.Assembly_real/*sed*scaffolds.fasta))\n",
    "FILENAME=${FILES[$SLURM_ARRAY_TASK_ID]}\n",
    "export SAMPLE=$(echo ${FILENAME} | sed -e 's/\\/home\\/htee663\\/ga02676\\/Waiwera_project\\/2.Assembly_real\\/\\(.*\\).scaffolds.fasta/\\1/g')    \n",
    "\n",
    "module load BBMap/37.93-gimkl-2017a\n",
    "\n",
    "#sediments\n",
    "dedupe.sh in=${FILENAME} out=dedupe.${SAMPLE}.scaffolds.fasta threads=24 minidentity=100\n",
    "\n",
    "#water\n",
    "cat *Filter.scaffolds.fasta > all.Filter.scaffolds.fasta\n",
    "dedupe.sh in=all.Filter.scaffolds.fasta out=dedupe.all.Filter.scaffolds.fasta threads=24 minidentity=100\n",
    "\n",
    "    \n",
    "# 4. mapping \n",
    "#job=\n",
    "mkdir 4-wts-waiwera-map-count\n",
    "    \n",
    "# 4.1 Water column slurm script \n",
    "\n",
    "TMPDIR=/nesi/nobackup/ga02676/Waiwera_project/szetest/tmp\n",
    "#mkdir -p $TMPDIR\n",
    "export TMPDIR\n",
    "\n",
    "export FILES=($(ls -1 3.trimmomatic/new.t30.q.mrna.RNA-F*_R1.fastq))\n",
    "FILENAME=${FILES[$SLURM_ARRAY_TASK_ID]}\n",
    "#Trimmed read directory:\n",
    "export DIR1=$(echo ${FILENAME} | sed -e 's/\\(3.trimmomatic\\)\\/.*/\\1/g')\n",
    "#Sample name:\n",
    "export SAMPLE=$(echo ${FILENAME} | sed -e 's/3.trimmomatic\\/new.t30.q.mrna.RNA-\\(.*\\)_R.*/\\1/g')\n",
    "#Contigs file name:\n",
    "export CONTIGS=$(echo /home/htee663/nobackupga02676/szetest/201023_dedupe/dedupe.all.filter.fasta)\n",
    "#Output directories:\n",
    "export DIR2=$(echo 4-wts-waiwera-map-count)\n",
    "\n",
    "\n",
    "## Map WTS read files to contigs:\n",
    "module load BBMap/37.93-gimkl-2017a\n",
    "\n",
    "srun bbmap.sh -Xmx160g in1=${DIR1}/new.t30.q.mrna.RNA-${SAMPLE}_R1.fastq \\\n",
    "  in2=${DIR1}/new.t30.q.mrna.RNA-${SAMPLE}_R2.fastq   \\\n",
    "    ref=${CONTIGS} t=36 nodisk  out=${DIR2}/${SAMPLE}.sam &> bbmap.rna.filter.${SAMPLE}.log\n",
    "\n",
    " grep -v '*' ${DIR2}/${SAMPLE}.sam > ${DIR2}/mapped.${SAMPLE}.sam\n",
    "\n",
    "module load SAMtools/1.8-gimkl-2017a\n",
    "\n",
    "srun samtools view -uS ${DIR2}/mapped.${SAMPLE}.sam | \\\n",
    "     samtools sort -O bam -o ${DIR2}/mapped.${SAMPLE}.bam\n",
    "\n",
    "    \n",
    "# 4.2 sediments slurm script \n",
    "\n",
    "\n",
    "TMPDIR=/nesi/nobackup/ga02676/Waiwera_project/szetest/tmp\n",
    "mkdir -p $TMPDIR\n",
    "export TMPDIR\n",
    "\n",
    "export FILES=($(ls -1 3.trimmomatic/new.t30.q.mrna.RNA-S*_R1.fastq))\n",
    "FILENAME=${FILES[$SLURM_ARRAY_TASK_ID]}\n",
    "#Trimmed read directory:\n",
    "export DIR1=$(echo ${FILENAME} | sed -e 's/\\(3.trimmomatic\\)\\/.*/\\1/g')\n",
    "#Sample name:\n",
    "export SAMPLE=$(echo ${FILENAME} | sed -e 's/3.trimmomatic\\/new.t30.q.mrna.RNA-\\(.*\\)_R.*/\\1/g')\n",
    "export NAME=$(echo ${FILENAME} | sed -e 's/3.trimmomatic\\/\\(.*\\)-\\(Sed[1-3]\\)_R.*/\\2/g')\n",
    "export SITE=$(echo ${FILENAME} | sed -e 's/3.trimmomatic\\/new.t30.q.mrna.*RNA-\\(S[1-7]\\)-Sed[1-3]_R.*/\\1/g')\n",
    "#Output directories:\n",
    "export DIR2=$(echo 4-wts-waiwera-map-count)\n",
    "export DIR3=$(echo /home/htee663/nobackupga02676/szetest/201023_dedupe/)\n",
    "\n",
    "## Map WTS read files to contigs:\n",
    "module load BBMap/37.93-gimkl-2017a\n",
    "\n",
    "srun bbmap.sh -Xmx120g in1=${DIR1}/new.t30.q.mrna.RNA-${SITE}-${NAME}_R1.fastq \\\n",
    "        in2=${DIR1}/new.t30.q.mrna.RNA-${SITE}-${NAME}_R2.fastq  \\\n",
    "    ref=${DIR3}/dedupe.${SITE}.sediment.scaffolds.fasta t=24 nodisk out=${DIR2}/${SAMPLE}.sam overwrite=true &> bbmap.rna.sed.${SITE}.log\n",
    "\n",
    "grep -v '*' ${DIR2}/${SAMPLE}.sam > ${DIR2}/mapped.${SAMPLE}.sam\n",
    "\n",
    "module load SAMtools/1.8-gimkl-2017a\n",
    "\n",
    "srun samtools view -uS ${DIR2}/mapped.${SAMPLE}.sam | \\\n",
    "     samtools sort -O bam -o ${DIR2}/mapped.${SAMPLE}.bam\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 5. hmmsearch every assembled contigs to get trusted annotation using WGS data (hmmsearch_gene_ann.sh)\n",
    "\n",
    "# 5.1 prodigal slurm script\n",
    "module load prodigal/2.6.3-GCCcore-7.4.0\n",
    "for i in *fasta; do \n",
    "    export NAME=$(echo $i | sed -e 's/\\.scaffolds.fasta//g'); \n",
    "    prodigal -i $i -o $NAME.gff -a $NAME.faa -d $NAME.fna -m -p met -f gff; \n",
    "    done\n",
    "\n",
    "\n",
    "module load HMMER/3.2.1-gimkl-2018b\n",
    "\n",
    "# optional: find trusted cutoff score using reference database (swissprot, seeds file from fungene)\n",
    "hmmsearch --cpu  12 --tblout fam.rubisco_small.txt  ~/database/Databases/szetestdatabase/PF00101.hmm uniprot-pf00101-filtered-reviewed_yes.fasta\n",
    "hmmsearch --cpu  12 --tblout fam.rubisco_large.txt  ~/database/Databases/szetestdatabase/PF00016.hmm uniprot-pf00101-filtered-reviewed_yes.fasta\n",
    "hmmsearch --cpu  12 --tblout fam.amoA_comammox.txt  ~/database/Databases/szetestdatabase/amoA_comammox.hmm amoA_comammox.seeds\n",
    "hmmsearch --cpu  12 --tblout fam.amoA_AOB_like.txt  ~/database/Databases/szetestdatabase/amoA_AOB_like.hmm amoA_AOB_like.seeds\n",
    "hmmsearch --cpu  12 --tblout fam.amoA_AOA.txt  ~/database/Databases/szetestdatabase/amoA_AOA.hmm amoA_AOA.seeds\n",
    "hmmsearch --cpu  12 --tblout fam.amoA_AOB.txt  ~/database/Databases/szetestdatabase/amoA_AOB.hmm amoA_AOB.seeds\n",
    "hmmsearch --cpu  12  --tblout fam.kdpA.txt ~/database/Databases/szetestdatabase/PF03814.hmm uniprot-kdpa-filtered-reviewed_yes.fasta\n",
    "hmmsearch --cpu  12   --tblout fam.kdpc.txt ~/database/Databases/szetestdatabase/PF02669.hmm uniprot-kdpc-reviewed_yes.fasta\n",
    "hmmsearch --cpu  12 -T 110 --tblout fam.nhaD.txt ~/database/Databases/szetestdatabase/PF03660.hmm nhaD.fasta\n",
    "\n",
    "# 5.2 hmmsearch against the database (refer supplementary table for genes of interested and the cutoff score)\n",
    "for n in {sediment,Filter};do for i in {1..9};do hmmsearch --cpu  12  -T  63 --tblout fam.S$i.$n.mnhG.txt ~/database/Databases/szetestdatabase/TIGR01300.hmm S$i.$n.faa;done;done\n",
    "for n in {sediment,Filter};do for i in {1..9};do hmmsearch --cpu  12  -T  24 --tblout fam.S$i.$n.mnhF.txt ~/database/Databases/szetestdatabase/PF04066.hmm S$i.$n.faa;done;done\n",
    "for n in {sediment,Filter};do for i in {1..9};do hmmsearch --cpu  12  -T  21 --tblout fam.S$i.$n.mnhE.txt ~/database/Databases/szetestdatabase/PF01899.hmm S$i.$n.faa;done;done\n",
    "for n in {sediment,Filter};do for i in {1..9};do hmmsearch --cpu  12  -T  692 --tblout fam.S$i.$n.mnhD.txt ~/database/Databases/szetestdatabase/TIGR00944.hmm S$i.$n.faa;done;done\n",
    "for n in {sediment,Filter};do for i in {1..9};do hmmsearch --cpu  12  -T  162 --tblout fam.S$i.$n.mnhC.txt ~/database/Databases/szetestdatabase/TIGR00941.hmm S$i.$n.faa;done;done\n",
    "for n in {sediment,Filter};do for i in {1..9};do hmmsearch --cpu  12  -T  22 --tblout fam.S$i.$n.mnhB.txt ~/database/Databases/szetestdatabase/PF04039.hmm S$i.$n.faa;done;done\n",
    "for n in {sediment,Filter};do for i in {1..9};do hmmsearch --cpu  12  -T  920 --tblout fam.S$i.$n.mnhA.txt ~/database/Databases/szetestdatabase/TIGR00940.hmm S$i.$n.faa;done;done\n",
    "for n in {sediment,Filter};do for i in {1..9};do hmmsearch --cpu  12  -T  220 --tblout fam.S$i.$n.NhaA.txt ~/database/Databases/szetestdatabase/TIGR00773.hmm S$i.$n.faa;done;done\n",
    "for n in {sediment,Filter};do for i in {1..9};do hmmsearch --cpu  12  -T  713 --tblout fam.S$i.$n.NhaB.txt ~/database/Databases/szetestdatabase/TIGR00774.hmm S$i.$n.faa;done;done\n",
    "for n in {sediment,Filter};do for i in {1..9};do hmmsearch --cpu  12  -T  409 --tblout fam.S$i.$n.NhaD.txt ~/database/Databases/szetestdatabase/TIGR00775.hmm S$i.$n.faa;done;done\n",
    "for n in {sediment,Filter};do for i in {1..9};do hmmsearch --cpu  12  -T  26 --tblout fam.S$i.$n.Na_H_Exchanger.txt ~/database/Databases/szetestdatabase/PF00999.hmm S$i.$n.faa;done;done\n",
    "for n in {sediment,Filter};do for i in {1..9};do hmmsearch --cpu  12  -T  444 --tblout fam.S$i.$n.kup.txt ~/database/Databases/szetestdatabase/TIGR00794.hmm S$i.$n.faa;done;done\n",
    "for n in {sediment,Filter};do for i in {1..9};do hmmsearch --cpu  12  -T  407 --tblout fam.S$i.$n.bcct.txt ~/database/Databases/szetestdatabase/TIGR00842.hmm S$i.$n.faa;done;done\n",
    "for n in {sediment,Filter};do for i in {1..9};do hmmsearch --cpu  12  -T  21 --tblout fam.S$i.$n.proX.txt ~/database/Databases/szetestdatabase/PF04069.hmm S$i.$n.faa;done;done\n",
    "for n in {sediment,Filter};do for i in {1..9};do hmmsearch --cpu  12  -T  326 --tblout fam.S$i.$n.proV.txt ~/database/Databases/szetestdatabase/TIGR01186.hmm S$i.$n.faa;done;done\n",
    "for n in {sediment,Filter};do for i in {1..9};do hmmsearch --cpu  12  -T  27 --tblout fam.S$i.$n.TrkA-C.txt ~/database/Databases/szetestdatabase/PF02080.hmm S$i.$n.faa;done;done\n",
    "for n in {sediment,Filter};do for i in {1..9};do hmmsearch --cpu  12  -T  295 --tblout fam.S$i.$n.TIGR01941.txt ~/database/Databases/szetestdatabase/TIGR01941.hmm S$i.$n.faa;done;done\n",
    "for n in {sediment,Filter};do for i in {1..9};do hmmsearch --cpu  12  -T  31 --tblout fam.S$i.$n.PF02406.txt ~/database/Databases/szetestdatabase/PF02406.hmm S$i.$n.faa;done;done\n",
    "for n in {sediment,Filter};do for i in {1..9};do hmmsearch --cpu  12  -T  24 --tblout fam.S$i.$n.PsaA_PsaB.txt ~/database/Databases/szetestdatabase/PF00223.hmm S$i.$n.faa;done;done\n",
    "for n in {sediment,Filter};do for i in {1..9};do hmmsearch --cpu  12  -T  48 --tblout fam.S$i.$n.PsaL.txt ~/database/Databases/szetestdatabase/PF02605.hmm S$i.$n.faa;done;done\n",
    "for n in {sediment,Filter};do for i in {1..9};do hmmsearch --cpu  12  -T  24 --tblout fam.S$i.$n.psaF.txt ~/database/Databases/szetestdatabase/PF02507.hmm S$i.$n.faa;done;done\n",
    "for n in {sediment,Filter};do for i in {1..9};do hmmsearch --cpu  12  -T  21 --tblout fam.S$i.$n.psbI.txt ~/database/Databases/szetestdatabase/PF02532.hmm S$i.$n.faa;done;done\n",
    "for n in {sediment,Filter};do for i in {1..9};do hmmsearch --cpu  12  -T  596 --tblout fam.S$i.$n.psbA.txt ~/database/Databases/szetestdatabase/TIGR01151.hmm S$i.$n.faa;done;done\n",
    "for n in {sediment,Filter};do for i in {1..9};do hmmsearch --cpu  12  -T  40 --tblout fam.S$i.$n.psbZ.txt ~/database/Databases/szetestdatabase/TIGR03043.hmm S$i.$n.faa;done;done\n",
    "for n in {sediment,Filter};do for i in {1..9};do hmmsearch --cpu  12  -T  25 --tblout fam.S$i.$n.PII.txt ~/database/Databases/szetestdatabase/PF00421.hmm S$i.$n.faa;done;done\n",
    "for n in {sediment,Filter};do for i in {1..9};do hmmsearch --cpu  12  -T  606 --tblout fam.S$i.$n.bchl.txt ~/database/Databases/szetestdatabase/PF02327.hmm S$i.$n.faa;done;done\n",
    "for n in {sediment,Filter};do for i in {1..9};do hmmsearch --cpu  12  -T  21 --tblout fam.S$i.$n.LHC.txt ~/database/Databases/szetestdatabase/PF00556.hmm S$i.$n.faa;done;done\n",
    "for n in {sediment,Filter};do for i in {1..9};do hmmsearch --cpu  12  -T  24 --tblout fam.S$i.$n.cellulase.txt ~/database/Databases/szetestdatabase/PF00150.hmm S$i.$n.faa;done;done\n",
    "r n in {sediment,Filter};do for i in {1..9};do hmmsearch --cpu  12  -T  729 --tblout fam.S$i.$n.cphA.txt ~/database/Databases/szetestdatabase/TIGR02068.hmm S$i.$n.faa;done;done\n",
    "for n in {sediment,Filter};do for i in {1..9};do hmmsearch --cpu  12  -T  126 --tblout fam.S$i.$n.cphB.txt ~/database/Databases/szetestdatabase/TIGR02069.hmm S$i.$n.faa;done;done\n",
    "for n in {sediment,Filter};do for i in {1..9};do hmmsearch --cpu  12  -T  527 --tblout fam.S$i.$n.kdpB.txt ~/database/Databases/szetestdatabase/TIGR01497.hmm S$i.$n.faa;done;done\n",
    "for n in {sediment,Filter};do for i in {1..9};do hmmsearch --cpu  12  -T  110 --tblout fam.S$i.$n.kdpA.txt ~/database/Databases/szetestdatabase/PF03814.hmm S$i.$n.faa;done;done\n",
    "for n in {sediment,Filter};do for i in {1..9};do hmmsearch --cpu  12  -T  110 --tblout fam.S$i.$n.kdpC.txt ~/database/Databases/szetestdatabase/PF02669.hmm S$i.$n.faa;done;done\n",
    "for n in {sediment,Filter};do for i in {1..9};do hmmsearch --cpu  12  -T  266 --tblout fam.S$i.$n.pufL.txt ~/database/Databases/szetestdatabase/TIGR01157.hmm S$i.$n.faa;done;done\n",
    "for n in {sediment,Filter};do for i in {1..9};do hmmsearch --cpu  12  -T  364 --tblout fam.S$i.$n.pufM.txt ~/database/Databases/szetestdatabase/TIGR01115.hmm S$i.$n.faa;done;done\n",
    "for n in {sediment,Filter};do for i in {1..9};do hmmsearch --cpu  12 -T 627.4 --tblout fam.S$i.$n.amoA_comammox.txt  ~/database/Databases/szetestdatabase/amoA_comammox.hmm S$i.$n.faa;done;done\n",
    "for n in {sediment,Filter};do for i in {1..9};do hmmsearch --cpu  12 -T 442 --tblout fam.S$i.$n.amoA_AOB_like.txt  ~/database/Databases/szetestdatabase/amoA_AOB_like.hmm S$i.$n.faa;done;done\n",
    "for n in {sediment,Filter};do for i in {1..9};do hmmsearch --cpu  12 -T 400 --tblout fam.S$i.$n.amoA_AOA.txt  ~/database/Databases/szetestdatabase/amoA_AOA.hmm S$i.$n.faa;done;done\n",
    "for n in {sediment,Filter};do for i in {1..9};do hmmsearch --cpu  12 -T 428.4 --tblout fam.S$i.$n.amoA_AOB.txt  ~/database/Databases/szetestdatabase/amoA_AOB.hmm S$i.$n.faa;done;done\n",
    "\n",
    "# 5.3 get the list of node name with functional genes of interested and the contig fasta sequences\n",
    "module load seqtk/1.2-gimkl-2017a\n",
    "for i in {1..9};do grep 'NODE' fam.S${i}.Filter* |sed  's/fam.*Filter.//g;s/.txt//g;s/:NODE/\\tNODE/g' |awk '{print $2,$1,$3,$4,$5,$6,$7}'>>all.fam.filter.S$i.txt;done\n",
    "for i in {1..9};do grep 'NODE' fam.S${i}.sed* |sed  's/fam.*Filter.//g;s/.txt//g;s/:NODE/\\tNODE/g' |awk '{print $2,$1,$3,$4,$5,$6,$7}'>>all.fam.sed.S$i.txt;done\n",
    "for i in {1..9};do grep 'NODE' all.fam.sed.S$i.txt |sed  's/ .*//g' >>list.fam.sed.S$i.txt;done\n",
    "for i in {1..9};do grep 'NODE' all.fam.filter.S$i.txt |sed  's/ .*//g' >>list.fam.filter.S$i.txt;done\n",
    "for i in {1..9};do seqtk subseq ../S$i.Filter.fna list.fam.filter.S$i.txt > all.fam.filter.S$i.fasta;done\n",
    "for i in {1..9};do seqtk subseq ../S$i.sediment.fna list.fam.sed.S$i.txt > all.fam.sed.S$i.fasta;done\n",
    "for i in {1..9};do grep 'NODE' all.fam.sed.S$i.txt |sed  's/ .*//g;s/\\(cov_[0-9]*.[0-9]*\\)_.*/\\1/g' >>list.fam.scaffold.sed.S$i.txt;done\n",
    "for i in {1..9};do grep 'NODE' all.fam.filter.S$i.txt |sed  's/ .*//g;s/\\(cov_[0-9]*.[0-9]*\\)_.*/\\1/g' >>list.fam.scaffold.filter.S$i.txt;done\n",
    "for i in list.fam.scaffold.*;do less $i|sort -u > sort.$i;done\n",
    "for i in {1..9};do seqtk subseq ~/ga02676/Waiwera_project/2.Assembly_real/S$i.Filter.scaffolds.fasta sort.list.fam.scaffold.filter.S$i.txt > all.fam.scaffold.filter.S$i.fasta;done\n",
    "for i in {1..9};do seqtk subseq ~/ga02676/Waiwera_project/2.Assembly_real/S$i.sediment.scaffolds.fasta sort.list.fam.scaffold.sed.S$i.txt > all.fam.scaffold.sed.S$i.fasta;done\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 6. Use kraken to identify contig taxonomy \n",
    "\n",
    "# get all the scaffolds\n",
    "cat S*sediment.scaffolds.fasta >>all.sed.scaffolds.fasta\n",
    "cat S*Filter.scaffolds.fasta >>all.filter.scaffolds.fasta\n",
    "\n",
    "# get all the ORF\n",
    "cat S*sediment.faa >>all.sed.faa\n",
    "cat S*Filter.faa >>all.filter.faa\n",
    "cat all.filter.faa all.sed.faa >>all.scaffold.faa\n",
    "\n",
    "#get annotated gene name and node name\n",
    "grep 'NODE' fam.*.Filter.* >> fam.all.filter.list.txt\n",
    "sed -i 's/fam.*Filter.//g;s/.txt//g;s/:NODE/\\tNODE/g' fam.all.filter.list.txt\n",
    "awk '{print $2,$1,$3,$4,$5,$6,$7}' fam.all.filter.list.txt >fam.all.gene.filter.list.txt\n",
    "rm fam.all.filter.list.txt\n",
    "\n",
    "grep 'NODE' fam.*.sediment.* >> fam.all.sed.list.txt\n",
    "sed -i 's/fam.*.sediment.//g;s/.txt//g;s/:NODE/\\tNODE/g' fam.all.sed.list.txt\n",
    "awk '{print $2,$1,$3,$4,$5,$6,$7}' fam.all.sed.list.txt >fam.all.gene.sed.list.txt\n",
    "rm fam.all.sed.list.txt\n",
    "\n",
    "#get the fasta sequence of gene of interested\n",
    "seqtk subseq ../all.sed.faa list.faa.all.sed.fam.txt > all.sed.fam.faa\n",
    "seqtk subseq ../all.filter.faa list.faa.all.filter.fam.txt > all.fil.fam.faa\n",
    "\n",
    "#get the fasta sequence of contig of interested \n",
    "seqtk subseq all.sed.scaffolds.fasta list.all.sed.fam.txt > all.sed.fam.fasta\n",
    "seqtk subseq all.filter.scaffolds.fasta list.all.filter.fam.txt > all.fil.fam.fasta\n",
    "\n",
    "\n",
    "# use kraken to identify taxa of the contig/gene of interest\n",
    "module load Kraken2/2.0.8-beta-gimkl-2018b\n",
    "\n",
    "kraken2 --db  ~/database/Databases/Kraken2/ --threads 16 --confidence 0.8 --report taxa_all.faa.sed.fam.txt --quick --use-mpa-style --use-name all.sed.fam.faa >taxa_all.faa.sed.fam.txt\n",
    "kraken2 --db  ~/database/Databases/Kraken2/ --threads 16 --confidence 0.8 --report taxa_all.faa.fil.fam.txt --quick --use-mpa-style --use-name all.fil.fam.faa >taxa_all.faa.fil.fam.txt\n",
    "\n",
    "kraken2 --db  ~/database/Databases/Kraken2/ --threads 16 --confidence 0.8 --report taxa_all.sed.fam.txt --quick --use-mpa-style --use-name all.sed.fam.fasta >taxa_all.sed.fam.txt\n",
    "kraken2 --db  ~/database/Databases/Kraken2/ --threads 16 --confidence 0.8 --report taxa_all.fil.fam.txt --quick --use-mpa-style --use-name all.fil.fam.fasta >taxa_all.fim.fam.txt\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 7. count reads with Featurecounts (subread packages)\n",
    "\n",
    "grep '>' all.filter.faa> list.all.filter.txt\n",
    "sed -i 's/>//g;s/#/\\t/g' list.all.filter.txt\n",
    "awk '{print $1,$1,$2,$3,$4}' list.all.filter.txt>new.list.all.filter.txt\n",
    "sed -i 's/_[0-9]*//12' new.list.all.filter.txt\n",
    "rm list.all.filter.txt\n",
    "echo 'GeneID Chr Start End Strand Feature' >edit.new.list.all.filter.gtf\n",
    "cat new.list.all.filter.txt>>edit.new.list.all.filter.gtf\n",
    "sed -i 's/ /\\t/g' edit.new.list.all.filter.gtf\n",
    "sed -i 's/\\(.*\\t.*\\t.*\\t.*\\t\\)1/\\1+/g' edit.new.list.all.filter.gtf\n",
    "sed -i 's/\\(.*\\t.*\\t.*\\t.*\\t\\)-1/\\1-/g' edit.new.list.all.filter.gtf\n",
    "sed 's/\\([+,-]\\)/\\1\\tCDS/g' edit.new.list.all.fil.gtf >test.edit.new.list.all.fil.gtf\n",
    "\n",
    "# sediment do individually due to overlap node name\n",
    "for i in {1..9};do grep '>' dedupe.S$i.sediment.faa > list.S$i.Sed.txt;done\n",
    "for i in {1..9};do sed -i 's/>//g;s/#/\\t/g' list.S$i.Sed.txt;done\n",
    "for i in {1..9};do awk '{print $1,$1,$2,$3,$4}' list.S${i}.Sed.txt >new.list.S${i}.sed.txt;done\n",
    "for i in {1..9};do sed -i 's/_[0-9]*//12' new.list.S${i}.sed.txt;done\n",
    "for i in {1..9};do echo 'GeneID Chr Start End Strand Feature' >edit.new.list.S$i.sed.gtf;done\n",
    "for i in {1..9};do sed 's/_[0-9]*//12' new.list.S$i.sed.txt >>edit.new.list.S$i.sed.gtf;done\n",
    "for i in {1..9};do sed -i 's/ /\\t/g;s/\\(.*\\t.*\\t.*\\t.*\\t\\)1/\\1+/g;s/\\(.*\\t.*\\t.*\\t.*\\t\\)-1/\\1-/g' edit.new.list.S$i.sed.gtf;done\n",
    "for i in {1..9};do sed 's/\\([+,-]\\)/\\1\\tCDS/g' edit.new.list.S$i.sed.gtf >test.edit.new.list.S$i.sed.gtf;done\n",
    "\n",
    "# example format required for featurecount input gtf file \n",
    "GeneID\tChr\tStart\tEnd\tStrand\n",
    "497097\tchr1\t3204563\t3207049\t-\n",
    "497097\tchr1\t3411783\t3411982\t-\n",
    "497097\tchr1\t3660633\t3661579\t-\n",
    "...\n",
    "\n",
    "#count number of ORF in each file\n",
    "for i in test.edit.new.list*;do echo $i;less $i|wc -l;done                           \n",
    "\n",
    "#get unique features from annotation file (make sure no repeat)\n",
    "for i in test.edit.new.list.S*;do less $i | awk '!a[$1]++' >01.$i;done\n",
    "\n",
    "#RNA\n",
    "ml Subread/2.0.0-GCC-9.2.0\n",
    "\n",
    "featureCounts -g GeneID -F SAF -t CDS -a test.edit.new.list.all.filter.gtf  \\\n",
    "-T 12 -o 3.annotation/featurecount_wts_all_filter.txt 4-wts-waiwera-map-count/mapped.Filt*bam\n",
    "awk -F \"\\t\" '{if(($7!=0) || ($8!=0) || ($9!=0)|| ($10!=0)|| ($11!=0)|| ($12!=0)|| ($13!=0)|| ($14!=0)|| ($15!=0)) {print}}' \\\n",
    "    3.annotation/featurecount_wts_all_filter.txt >3.annotation/edited.featurecount_wts_all_filter.txt\n",
    "\n",
    "for i in {1..7};do \n",
    "    featureCounts -p -g GeneID -t CDS -F SAF -a 01.test.edit.new.list.S${i}.sed.gtf \\\n",
    "    -T 24 -o 3.annotation/featurecount_wts_s${i}_sed.txt 4-wts-waiwera-map-count/mapped.S${i}-Sed?.bam;\n",
    "    done\n",
    "for i in {1..7};do \n",
    "    awk '{if($7!=0 || $8!=0 || $9!=0) print}' 3.annotation/featurecount_wts_s${i}_sed.txt \\\n",
    "    >3.annotation/edit.featurecount_wts_s${i}_sed.txt;done\n",
    "\n",
    "#DNA\n",
    "featureCounts -g GeneID -F SAF -t CDS -a 2-wgs-waiwera-map-count/test.edit.new.list.all.filter.gtf \\\n",
    "    -T 12 -o 3.annotation/featurecount_wgs_all_filter.txt 2-wgs-waiwera-map-count/mapped.Filt.S*bam\n",
    "awk -F \"\\t\" '{if(($7!=0) || ($8!=0) || ($9!=0)|| ($10!=0)|| ($11!=0)|| ($12!=0)|| ($13!=0)|| ($14!=0)|| ($15!=0)) {print}}' \\\n",
    "    3.annotation/featurecount_wgs_all_filter.txt >3.annotation/edited.featurecount_wgs_all_filter.txt\n",
    "\n",
    "for i in {1..9};do \n",
    "    featureCounts -p -g GeneID -t CDS -F SAF -a 2-wgs-waiwera-map-count/01.test.edit.new.list.S${i}.sed.gtf  \\\n",
    "    -T 24 -o 3.annotation/featurecount_wgs_s${i}_sed.txt 2-wgs-waiwera-map-count/mapped.Sed.S${i}R?.bam;done\n",
    "\n",
    "for i in {1..9};do \n",
    "    awk '{if($7!=0 || $8!=0 || $9!=0) print}' 3.annotation/featurecount_wgs_s${i}_sed.txt \\\n",
    "    >3.annotation/edit.featurecount_wgs_s${i}_sed.txt;done\n",
    "    \n",
    "#add site name to the sample\n",
    "for f in s1 s2 s3 s4 s5 s6 s7 s8 s9; do sed \"s/$/\\t$f/\" edit.featurecount_wgs_$f\\_sed.txt >nedit.featurecount_wgs_$f\\_sed.txt; done\n",
    "for f in s1 s2 s3 s4 s5 s6 s7; do sed \"s/$/\\t$f/\" edit.featurecount_wts_$f\\_sed.txt >nedit.featurecount_wts_$f\\_sed.txt; done\n",
    "for f in edited.featurecount_w*all_filter*; do sed \"s/$/\\tall filter/\" $f > n$f; done\n",
    "\n",
    "for i in 1 2 3 4 5 6 7 8 9;do sed -i \"1d; 0,/s${i}/s//site/\" \"nedit.featurecount_wgs_s${i}_sed.txt\" ;done\n",
    "for i in 1 2 3 4 5 6 7;do sed -i \"1d; 0,/s${i}/s//site/\" \"nedit.featurecount_wts_s${i}_sed.txt\" ;done\n",
    "for i in nedited.featurecount_w*all_filter*; do sed -i \"1d; 0,/all filter/s//site/\" \"$i\" ;done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 8. TPM\n",
    "# 8.1 Get mapped mRNA library size (in bash)\n",
    "\n",
    "# saved in a file called \"CountMappedReads.sh\"\n",
    "\n",
    "#! /bin/bash\n",
    "\n",
    "ml SAMtools/1.10-GCC-9.2.0\n",
    "# file header\n",
    "echo \"bam_file total mapped unmapped\"\n",
    "\n",
    "# loop to count reads in all files and add results to a table\n",
    "for sam_file in *.sam\n",
    "do\n",
    "total=$(samtools view -c $sam_file)\n",
    "mapped=$(samtools view -c -F 4 $sam_file)\n",
    "unmapped=$(samtools view -c -f 4 $sam_file)\n",
    "echo \"$sam_file $total $mapped $unmapped\"\n",
    "done\n",
    "\n",
    "#then run in bash\n",
    "./CountMappedReads.sh\n",
    " \n",
    "\n",
    "# 8.2 Get mapped DNA library size\n",
    "#get from DNA read mapping files using the script as above\n",
    "\n",
    "# 8.3 read in count files using R (gene_count.r)\n",
    "#calculate tpm\n",
    "library(plyr)\n",
    "library(readr)\n",
    "library(\"dplyr\")\n",
    "library(\"gplots\")\n",
    "\n",
    "wts_filter<-read_delim(\"nedited.featurecount_wts_all_filter.txt\",delim=\"\\t\")\n",
    "myfiles = list.files(path=\".\", pattern=\"nedit.featurecount_wts_s[1-7]_sed.txt\", full.names=TRUE)\n",
    "wts_sed = ldply(myfiles, read_delim, delim=\"\\t\")\n",
    "wts_sed=wts_sed[, c(1:6,10,7:9,11:28)]\n",
    "\n",
    "wgs_filter<-read_delim(\"nedited.featurecount_wgs_all_filter.txt\",delim=\"\\t\")\n",
    "myfiles = list.files(path=\".\", pattern=\"nedit.featurecount_wgs_s[1-9]_sed.txt\", full.names=TRUE)\n",
    "wgs_sed = ldply(myfiles, read_delim, delim=\"\\t\")\n",
    "wgs_sed=wgs_sed[, c(1:6,10,7:9,11:34)]\n",
    "\n",
    "lib.size_mappedRNA <- read_delim(\"mapped_mRNA_libsize.txt\",delim = \"\\t\")\n",
    "lib.size_mappedDNA <- read_delim(\"mapped_DNA_libsize.txt\",delim = \"\\t\")\n",
    "\n",
    "\n",
    "count_wts_filter<- wts_filter[,7:15]\n",
    "count_wts_sed<- wts_sed[,8:28]\n",
    "count_wts_filter[,1:9]<-lapply(count_wts_filter[,1:9],as.numeric)\n",
    "count_wts_sed[,1:21]<-lapply(count_wts_sed[,1:21],as.numeric)\n",
    "\n",
    "count_wgs_filter<- wgs_filter[,7:15]\n",
    "count_wgs_sed<- wgs_sed[,8:34]\n",
    "count_wgs_filter[,1:9]<-lapply(count_wgs_filter[,1:9],as.numeric)\n",
    "count_wgs_sed[,1:27]<-lapply(count_wgs_sed[,1:27],as.numeric)\n",
    "\n",
    "\n",
    "gene_fam_filter<-read_delim(\"fam.all.gene.filter.list.txt\",delim=\" \",col_names = FALSE)\n",
    "gene_fam_filter<-gene_fam_filter[,c(1,2,4,5,6,7)]\n",
    "\n",
    "gene_fam_sed<-read_delim(\"fam.all.gene.sed.list.txt\",delim=\" \",col_names = FALSE)\n",
    "gene_fam_sed<-gene_fam_sed[,c(1,2,4,5,6,7)]\n",
    "\n",
    "tpm <- function(counts, lengths) {\n",
    "  rate <- counts / lengths\n",
    "  rate / sum(rate) * 1e6\n",
    "}\n",
    "\n",
    "\n",
    "totalrate=as.data.frame(colSums(count_wts_filter/wts_filter$Length))\n",
    "totalratesed=as.data.frame(colSums(count_wts_sed/wts_sed$Length,na.rm= T))\n",
    "write_delim(totalrate,\"totalrate_wts_fil.txt\")\n",
    "write_delim(totalratesed,\"totalrate_wts_sed.txt\")\n",
    "\n",
    "gtotalrate=as.data.frame(colSums(count_wgs_filter/wgs_filter$Length,na.rm= T))\n",
    "gtotalratesed=as.data.frame(colSums(count_wgs_sed/wgs_sed$Length,na.rm= T))\n",
    "write_delim(gtotalrate,\"totalrate_wgs_fil.txt\")\n",
    "write_delim(gtotalratesed,\"totalrate_wgs_sed.txt\")\n",
    "\n",
    "tpm_wts_fil=(wts_filter[,7:15]/wts_filter$Length)/(totalrate[col(wts_filter[,7:15])])*1e6\n",
    "tpm_wts_sed=(wts_sed[,8:28]/wts_sed$Length)/(totalratesed[col(wts_sed[,8:28])])*1e6\n",
    "\n",
    "tpm_wgs_fil=(wgs_filter[,7:15]/wgs_filter$Length)/(gtotalrate[col(wgs_filter[,7:15])])*1e6\n",
    "tpm_wgs_sed=(wgs_sed[,8:34]/wgs_sed$Length)/(gtotalratesed[col(wgs_sed[,8:34])])*1e6\n",
    "\n",
    "colSums(tpm_wts_fil, na.rm= T)\n",
    "colSums(tpm_wts_sed, na.rm= T)\n",
    "colSums(tpm_wgs_fil, na.rm= T)\n",
    "colSums(tpm_wgs_sed, na.rm= T)\n",
    "\n",
    "tpm_wts_fil1=cbind(wts_filter[c(1:6,16)],tpm_wts_fil)\n",
    "tpm_wts_sed1=cbind(wts_sed[1:7],tpm_wts_sed)\n",
    "tpm_wgs_fil1=cbind(wgs_filter[c(1:6,16)],tpm_wgs_fil)\n",
    "tpm_wgs_sed1=cbind(wgs_sed[1:7],tpm_wgs_sed)\n",
    "\n",
    "\n",
    "\n",
    "combgene_wts_filter<-left_join(gene_fam_filter,tpm_wts_fil1,by=c(\"X1\"=\"Geneid\"))%>% filter(Length!=\"NA\")\n",
    "combgene_wts_sed<-left_join(gene_fam_sed,tpm_wts_sed1,by=c(\"X1\"=\"Geneid\"))%>% filter(Length!=\"NA\")\n",
    "combgene_wgs_filter<-left_join(gene_fam_filter,tpm_wgs_fil1,by=c(\"X1\"=\"Geneid\"))%>% filter(Length!=\"NA\")\n",
    "combgene_wgs_sed<-left_join(gene_fam_sed,tpm_wgs_sed1,by=c(\"X1\"=\"Geneid\"))%>% filter(Length!=\"NA\")\n",
    "write_delim(combgene_wgs_filter,\"combgene_wgs_filter.txt\")\n",
    "write_delim(combgene_wts_filter,\"combgene_wts_filter.txt\")\n",
    "write_delim(combgene_wgs_sed,\"combgene_wgs_sed.txt\")\n",
    "write_delim(combgene_wts_sed,\"combgene_wts_sed.txt\")\n",
    "\n",
    "\n",
    "#combgene_wgs_filter=read.delim(file=\"combgene_wgs_filter.txt\", sep=\"\\t\", header=T)\n",
    "#combgene_wgs_sed=read.delim(file=\"combgene_wgs_sed.txt\", sep=\"\\t\", header=T)\n",
    "\n",
    "#filtered out genes that are duplicated/not interested\n",
    "edit_gene_wgs_filter<-combgene_wgs_filter %>% filter(X2!=\"NA_sulfate\",X2!=\"GROEL\",X2!=\"pfk-eu\",X2!=\"pfk-prok\",X2!=\"rubisco_small\", X2!=\"rubisco_large\",X2!=\"TIGR02891\",X2!=\"TrkA-N\",X2!=\"Na_H_Exchanger\",X2!=\"TIGR00455\",X2!=\"TIGR00780\",X2!=\"TIGR03479\",X2!=\"TIGR03478\",X2!=\"PF01654\",X2!=\"TIGR02416\",X2!=\"TIGR02866\",X2!=\"TIGR02891\",X2!=\"TIGR00781\",X2!=\"TIGR00782\",X2!=\"TIGR02882\",X2!=\"TIGR01432\",X2!=\"PF01654\",X2!=\"TIGR00203\",X2!=\"TIGR01433\",X2!=\"TIGR01473\",X2!=\"TIGR02847\",X2!=\"TIGR00193\",X2!=\"TIGR01792\")\n",
    "edit_gene_wgs_sed<-combgene_wgs_sed %>% filter(X2!=\"NA_sulfate\",X2!=\"GROEL\",X2!=\"pfk-eu\",X2!=\"pfk-prok\",X2!=\"rubisco_small\", X2!=\"rubisco_large\",X2!=\"TIGR02891\",X2!=\"TrkA-N\",X2!=\"Na_H_Exchanger\",X2!=\"TIGR00455\",X2!=\"TIGR00780\",X2!=\"TIGR03479\",X2!=\"TIGR03478\",X2!=\"PF01654\",X2!=\"TIGR02416\",X2!=\"TIGR02866\",X2!=\"TIGR02891\",X2!=\"TIGR00781\",X2!=\"TIGR00782\",X2!=\"TIGR02882\",X2!=\"TIGR01432\",X2!=\"PF01654\",X2!=\"TIGR00203\",X2!=\"TIGR01433\",X2!=\"TIGR01473\",X2!=\"TIGR02847\",X2!=\"TIGR00193\",X2!=\"TIGR01792\")\n",
    "edit_gene_wts_sed<-combgene_wts_sed %>% filter(X2!=\"NA_sulfate\",X2!=\"GROEL\",X2!=\"pfk-eu\",X2!=\"pfk-prok\",X2!=\"rubisco_small\", X2!=\"rubisco_large\",X2!=\"TIGR02891\",X2!=\"TrkA-N\",X2!=\"Na_H_Exchanger\",X2!=\"TIGR00455\",X2!=\"TIGR00780\",X2!=\"TIGR03478\",X2!=\"TIGR03479\",X2!=\"PF01654\",X2!=\"TIGR02416\",X2!=\"TIGR02866\",X2!=\"TIGR02891\",X2!=\"TIGR00781\",X2!=\"TIGR00782\",X2!=\"TIGR02882\",X2!=\"TIGR01432\",X2!=\"PF01654\",X2!=\"TIGR00203\",X2!=\"TIGR01433\",X2!=\"TIGR01473\",X2!=\"TIGR02847\",X2!=\"TIGR00193\",X2!=\"TIGR01792\")\n",
    "edit_gene_wts_filter<-combgene_wts_filter %>% filter(X2!=\"NA_sulfate\",X2!=\"GROEL\",X2!=\"pfk-eu\",X2!=\"pfk-prok\",X2!=\"rubisco_small\", X2!=\"rubisco_large\",X2!=\"TIGR02891\",X2!=\"TrkA-N\",X2!=\"Na_H_Exchanger\",X2!=\"TIGR00455\",X2!=\"TIGR00780\",X2!=\"TIGR03478\",X2!=\"TIGR03479\",X2!=\"PF01654\",X2!=\"TIGR02416\",X2!=\"TIGR02866\",X2!=\"TIGR02891\",X2!=\"TIGR00781\",X2!=\"TIGR00782\",X2!=\"TIGR02882\",X2!=\"TIGR01432\",X2!=\"PF01654\",X2!=\"TIGR00203\",X2!=\"TIGR01433\",X2!=\"TIGR01473\",X2!=\"TIGR02847\",X2!=\"TIGR00193\",X2!=\"TIGR01792\")\n",
    "\n",
    "#count no occurance\n",
    "#see which node with more than 1 count\n",
    "n_occur <- data.frame(table(edit_gene_wgs_filter$X1))\n",
    "TEST2<-edit_gene_wgs_filter[edit_gene_wgs_filter$X1 %in% n_occur$Var1[n_occur$Freq > 1],]\n",
    "edit_gene_wgs_filter<-edit_gene_wgs_filter[order(-edit_gene_wgs_filter$X7,edit_gene_wgs_filter$X1),] %>% distinct(X1,.keep_all = TRUE)\n",
    "\n",
    "n_occur <- data.frame(table(edit_gene_wgs_sed$X1))\n",
    "TEST3<-edit_gene_wgs_sed[edit_gene_wgs_sed$X1 %in% n_occur$Var1[n_occur$Freq > 1],]\n",
    "edit_gene_wgs_sed<-edit_gene_wgs_sed[order(-edit_gene_wgs_sed$X7,edit_gene_wgs_sed$X1),] %>% distinct(X1,.keep_all = TRUE)\n",
    "\n",
    "n_occur <- data.frame(table(edit_gene_wts_filter$X1))\n",
    "TEST2<-edit_gene_wts_filter[edit_gene_wts_filter$X1 %in% n_occur$Var1[n_occur$Freq > 1],]\n",
    "edit_gene_wts_filter<-edit_gene_wts_filter[order(-edit_gene_wts_filter$X7,edit_gene_wts_filter$X1),] %>% distinct(X1,.keep_all = TRUE)\n",
    "\n",
    "n_occur <- data.frame(table(edit_gene_wts_sed$X1))\n",
    "TEST3<-edit_gene_wts_sed[edit_gene_wts_sed$X1 %in% n_occur$Var1[n_occur$Freq > 1],]\n",
    "edit_gene_wts_sed<-edit_gene_wts_sed[order(-edit_gene_wts_sed$X7,edit_gene_wts_sed$X1),] %>% distinct(X1,.keep_all = TRUE)\n",
    "\n",
    "\n",
    "taxa_wts_all<-full_join(edit_gene_wts_filter,edit_gene_wts_sed,by=c(\"X1\"=\"X1\"))\n",
    "taxa_wts_all[,22:32]<-lapply(taxa_wts_all[,22:32],as.character)\n",
    "taxa_wts_all[,2:12]<-lapply(taxa_wts_all[,2:12],as.character)\n",
    "v=which(is.na(taxa_wts_all$`4-wts-waiwera-map-count/mapped.Filt1.bam`)) %>% head(1)\n",
    "taxa_wts_all[1:v,22:32]<-taxa_wts_all[1:v,2:12]\n",
    "taxa_wts_all <- taxa_wts_all[ -c(2:12) ]\n",
    "taxa_wts_all<-taxa_wts_all[, c(1,11:21,2:10,22:42)]\n",
    "taxa_wts_all<- taxa_wts_all%>% filter(Length.y!=\"NA\")\n",
    "taxa_wts_all[,13:42][is.na(taxa_wts_all[,13:42])]<-0\n",
    "\n",
    "colnames(taxa_wts_all) <-c(\"node\",\"n1\",\"n2\",\"n3\",\"n4\",\"n5\",\"n6\",\"n7\",\"n8\",\"n9\",\"n10\",\"n11\",\"f1\",\"f2\",\"f3\",\"f4\",\"f5\",\"f6\",\"f7\",\"f8\",\"f9\",\"S1R1\",\"S1R2\",\"S1R3\",\"S2R1\",\"S2R2\",\"S2R3\",\"S3R1\",\"S3R2\",\"S3R3\",\"S4R1\",\"S4R2\",\"S4R3\",\"S5R1\",\"S5R2\",\"S5R3\",\"S6R1\",\"S6R2\",\"S6R3\",\"S7R1\",\"S7R2\",\"S7R3\")\n",
    "write_delim(taxa_wts_all,\"tpm_wts_all.txt\")\n",
    "\n",
    "taxa_wgs_all<-full_join(edit_gene_wgs_filter,edit_gene_wgs_sed,by=c(\"X1\"=\"X1\"))\n",
    "taxa_wgs_all[,22:32]<-lapply(taxa_wgs_all[,22:32],as.character)\n",
    "taxa_wgs_all[,2:12]<-lapply(taxa_wgs_all[,2:12],as.character)\n",
    "v=which(is.na(taxa_wgs_all$mapped.RNA_S1.Fil.bam)) %>% head(1)\n",
    "taxa_wgs_all[1:v,22:32]<-taxa_wgs_all[1:v,2:12]\n",
    "taxa_wgs_all <- taxa_wgs_all[ -c(2:12) ]\n",
    "taxa_wgs_all<-taxa_wgs_all[, c(1,11:21,2:10,22:48)]\n",
    "taxa_wgs_all<- taxa_wgs_all%>% filter(Length.y!=\"NA\")\n",
    "taxa_wgs_all[,13:48][is.na(taxa_wgs_all[,13:48])]<-0\n",
    "\n",
    "colnames(taxa_wgs_all) <-c(\"node\",\"n1\",\"n2\",\"n3\",\"n4\",\"n5\",\"n6\",\"n7\",\"n8\",\"n9\",\"n10\",\"n11\",\"f1\",\"f2\",\"f3\",\"f4\",\"f5\",\"f6\",\"f7\",\"f8\",\"f9\",\"S1R1\",\"S1R2\",\"S1R3\",\"S2R1\",\"S2R2\",\"S2R3\",\"S3R1\",\"S3R2\",\"S3R3\",\"S4R1\",\"S4R2\",\"S4R3\",\"S5R1\",\"S5R2\",\"S5R3\",\"S6R1\",\"S6R2\",\"S6R3\",\"S7R1\",\"S7R2\",\"S7R3\",\"S8R1\",\"S8R2\",\"S8R3\",\"S9R1\",\"S9R2\",\"S9R3\")\n",
    "write_delim(taxa_wgs_all,\"tpm_wgs_all.txt\")\n",
    "\n",
    "\n",
    "remove(wts_filter,wgs_filter,wts_sed,wgs_sed,count_wts_filter,count_wgs_filter,count_wts_sed,count_wgs_sed)\n",
    "\n",
    "save.image(file = 'tpm_wtswgs.RData')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 9. get taxonomy from all node from tpm (wgs wts) list\n",
    "#as there are duplicate nodenames within sediment samples, I have to retrieve the nodename individually from each sample and get their corresponding fasta files for blast to get their taxonomy\n",
    "for ((i=1; i<=7; i++)); do echo $(awk -F $'\\t' -v i=$i '{if($12~i) {print $1}}' tpm_wts_all.txt) | awk 'BEGIN { OFS = \"\\n\" } { $1=$1; print }' |sort -u  >nodelist.wts.s$i.txt ; done\n",
    "for ((i=1; i<=9; i++)); do echo $(awk -F $'\\t' -v i=$i '{if($12~i) {print $1}}' tpm_wgs_all.txt) | awk 'BEGIN { OFS = \"\\n\" } { $1=$1; print }' |sort -u >nodelist.wgs.s$i.txt ; done\n",
    "(awk -F $'\\t'  '{if($12~\"all\") print $1}' tpm_wts_all.txt) | awk 'BEGIN { OFS = \"\\n\" } { $1=$1; print }' |sort -u >nodelist.wts.allf.txt\n",
    "(awk -F $'\\t'  '{if($12~\"all\") print $1}' tpm_wgs_all.txt) | awk 'BEGIN { OFS = \"\\n\" } { $1=$1; print }'| sort -u >nodelist.wgs.allf.txt\n",
    "\n",
    "for i in {1..7};do comm nodelist.wts.s$i.txt nodelist.wgs.s$i.txt |sed 's/\\s//g' > nodelist.wgswts.s$i.txt;done\n",
    "\n",
    "comm nodelist.wts.allf.txt nodelist.wgs.allf.txt |sed 's/\\s//g' > nodelist.wgswts.allf.txt\n",
    "\n",
    "module load seqtk/1.2-gimkl-2017a\n",
    "seqtk subseq ../../../201023_dedupe/dedupe.all.filter.fasta.faa nodelist.wgswts.allf.txt > node.filter.fasta\n",
    "for i in {1..7}; do seqtk subseq ../../../201023_dedupe/dedupe.S$i.sediment.faa nodelist.wgswts.s$i.txt > node.s$i.fasta;done\n",
    "for i in {8..9}; do seqtk subseq ../../../201023_dedupe/dedupe.S$i.sediment.faa nodelist.wgs.s$i.txt > node.s$i.fasta;done\n",
    "\n",
    "#####NR DATABASE\n",
    "ml DIAMOND/2.0.6-GCC-9.2.0\n",
    "\n",
    "diamond makedb --in NCBI_20200213/nr.gz -d nr.dmnd \\ \n",
    "    --taxonmap prot.accession2taxid \\ \n",
    "    --taxonnodes taxdump/nodes.dmp \\ \n",
    "    --taxonnames taxdump/names.dmp\n",
    "\n",
    "for i in node.*fasta;do diamond blastp -p 30  --id 80 --query-cover 70 --taxon-exclude 77133,358574,12908,151659 \\\n",
    "    --sensitive --top 5 -c1 -d nr.dmnd -q $i -o $i.txt \\ \n",
    "    -t /tmp/ --outfmt 102\n",
    "\n",
    "#check how many with annotation\n",
    "for i in qe.dmnd_nr*txt;do awk '$2!=0 {print $0}' $i |wc -l;done\n",
    "for i in qe.dmnd_nr*txt;do awk '!seen[$1]++' $i  >uniq.$i; done\n",
    "header=\"node\\ttaxa\\te-value\"\n",
    "for file in uniq.qe.dmnd_nr_node.*;do sed -i -e \"1i\\ $header\" $file; done\n",
    "\n",
    "#combine multiple file into 1\n",
    "awk 'FNR==1{if (NR==1) print \"filename\", $0; next} {print FILENAME, $0}' uniq.qe.dmnd_nr_node.* >uniq.qe.dmnd_nr_all.txt\n",
    "sed -i '1d' uniq.qe.dmnd_nr_all.txt\n",
    "\n",
    "#get taxonomy\n",
    "ml Miniconda3/4.8.3\n",
    "python get_taxonomy_lineage.py -i uniq.qe.dmnd_nr_all.txt -o uniq.qe.taxa.dmnd_nr_all.txt\n",
    "sed -i 's/uniq.qe.dmnd_nr_node.\\(.*\\).fasta.txt /\\1\\t/g;s/filter/all filter/g' uniq.qe.taxa.dmnd_nr_all.txt\n",
    "\n",
    "#rename some of the taxonomy names\n",
    "sed -i 's/FCB group\\t//g;s/Bacteroidetes\\/Chlorobi group\\t//g;s/Cyanobacteria\\/Melainabacteria group\\t//g;s/Asgard group\\t//g' uniq.qe.taxa.dmnd_nr_all.txt\n",
    "sed -i 's/Terrabacteria group\\t//g;s/PVC group\\t//g;s/TACK group\\t//g;s/Nitrospinae\\/Tectomicrobia group\\t//g;s/delta\\/epsilon subdivisions\\t//g' uniq.qe.taxa.dmnd_nr_all.txt \n",
    "sed -i 's/Candidatus //g;s/candidate division //g;s/ bacterium//g' uniq.qe.taxa.dmnd_nr_all.txt\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# 10. match taxonomy to tpm data using R\n",
    "library(\"ggplot2\")\n",
    "library(\"dplyr\")\n",
    "library(\"readr\")\n",
    "library(\"tidyr\")\n",
    "library(\"stringr\")\n",
    "\n",
    "#add col name first\n",
    "taxa<-read.delim2(\"uniq.qe.taxa.dmnd_nr_all.txt\",sep=\"\\t\",header = TRUE) \n",
    "label<-read.delim2(\"label.txt\",sep=\"\\t\")\n",
    "\n",
    "taxa_wgs_all<-read.delim2(\"tpm_wgs_all.txt\",sep = \"\\t\")\n",
    "taxa_wts_all<-read.delim2(\"tpm_wts_all.txt\",sep = \"\\t\")\n",
    "\n",
    "taxa_wgs_all<-left_join(taxa_wgs_all,label,by=c(\"n1\"=\"X1\"))  \n",
    "taxa_wts_all<-left_join(taxa_wts_all,label,by=c(\"n1\"=\"X1\"))\n",
    "\n",
    "taxa_wgs_all$newnode=taxa_wgs_all$node\n",
    "taxa_wts_all$newnode=taxa_wts_all$node\n",
    "\n",
    "taxa_wgs_all$newnode=gsub('(cov_[0-9]*.[0-9]*)_.*', '\\\\1', taxa_wgs_all$newnode)\n",
    "taxa_wts_all$newnode=gsub('(cov_[0-9]*.[0-9]*)_.*', '\\\\1', taxa_wts_all$newnode)\n",
    "\n",
    "taxa_wgs_all<-left_join(taxa_wgs_all,taxa,by=c(\"node\"=\"node\", \"n11\"=\"sample\"))\n",
    "taxa_wts_all<-left_join(taxa_wts_all,taxa,by=c(\"node\"=\"node\", \"n11\"=\"sample\"))\n",
    "\n",
    "taxa_wgs_all[,54:71][is.na(taxa_wgs_all[,54:71])]<-\"\"\n",
    "taxa_wts_all[,48:65][is.na(taxa_wts_all[,48:65])]<-\"\"\n",
    "\n",
    "taxa_wgs_all[,54:71] <- lapply(taxa_wgs_all[,54:71], as.character)\n",
    "taxa_wts_all[,48:65]<- lapply(taxa_wts_all[,48:65], as.character)\n",
    "\n",
    "taxa_wgs_all[,54:71] <- lapply(taxa_wgs_all[,54:71], as.character)\n",
    "taxa_wts_all[,48:65]<- lapply(taxa_wts_all[,48:65], as.character)\n",
    "taxa_wgs_all<-taxa_wgs_all %>%mutate(t5 = case_when(t4 %in% c(\"unclassified Archaea\",\"environmental samples\",\"unclassified Bacteria\",\"uncultured Mediterranean phage\") ~ \"\",TRUE ~ as.character(t5)),\n",
    "                                     t3 = case_when(t3 %in% c(\"environmental samples\",\"unclassified Bacteria\",\"uncultured Mediterranean phage\") ~ \"\",TRUE ~ as.character(t3)),\n",
    "                                     t4 = case_when(t4 %in% c(\"unclassified Archaea\",\"environmental samples\",\"unclassified Bacteria\",\"uncultured Mediterranean phage\") ~ \"\",TRUE ~ as.character(t4)),\n",
    "                                     t5 = case_when(t5 %in% c(\"unclassified Archaea\",\"environmental samples\",\"unclassified Bacteria\",\"uncultured Mediterranean phage\") ~ \"\",TRUE ~ as.character(t5)))\n",
    "\n",
    "taxa_wts_all<-taxa_wts_all %>%mutate(t5 = case_when(t4 %in% c(\"unclassified Archaea\",\"environmental samples\",\"unclassified Bacteria\",\"uncultured Mediterranean phage\") ~ \"\",TRUE ~ as.character(t5)),\n",
    "                                     t3 = case_when(t3 %in% c(\"environmental samples\",\"unclassified Bacteria\",\"uncultured Mediterranean phage\") ~ \"\",TRUE ~ as.character(t3)),\n",
    "                                     t4 = case_when(t4 %in% c(\"unclassified Archaea\",\"environmental samples\",\"unclassified Bacteria\",\"uncultured Mediterranean phage\") ~ \"\",TRUE ~ as.character(t4)),\n",
    "                                     t5 = case_when(t5 %in% c(\"unclassified Archaea\",\"environmental samples\",\"unclassified Bacteria\",\"uncultured Mediterranean phage\") ~ \"\",TRUE ~ as.character(t5))) \n",
    "\n",
    "taxa_wts_all$t5<-str_replace(taxa_wts_all$t5,\"unclassified*\",\"\") \n",
    "taxa_wgs_all$t5<-str_replace(taxa_wgs_all$t5,\"unclassified*\",\"\") \n",
    "\n",
    "taxa_wgs_all$t3[taxa_wgs_all$t3==\"\"]<-\"unknown\"\n",
    "taxa_wts_all$t3[taxa_wts_all$t3==\"\"]<-\"unknown\"\n",
    "\n",
    "taxa_wgs_all<-taxa_wgs_all%>% unite(\"taxa\",t3:t18,remove = FALSE) \n",
    "taxa_wts_all<-taxa_wts_all%>% unite(\"taxa\",t3:t14,remove = FALSE) \n",
    "\n",
    "taxa_wgs_all<-taxa_wgs_all%>% unite(\"taxac\",t3:t5,remove = FALSE) \n",
    "taxa_wts_all<-taxa_wts_all%>% unite(\"taxac\",t3:t5,remove = FALSE) \n",
    "\n",
    "taxa_wgs_all<-taxa_wgs_all%>% unite(\"taxaorder\",t3:t6,remove = FALSE) \n",
    "taxa_wts_all<-taxa_wts_all%>% unite(\"taxaorder\",t3:t6,remove = FALSE) \n",
    "\n",
    "taxa_wgs_all<-taxa_wgs_all%>% unite(\"taxafam\",t3:t7,remove = FALSE) \n",
    "taxa_wts_all<-taxa_wts_all%>% unite(\"taxafam\",t3:t7,remove = FALSE) \n",
    "\n",
    "taxa_wgs_all<-taxa_wgs_all%>% unite(\"taxagen\",t3:t8,remove = FALSE) \n",
    "taxa_wts_all<-taxa_wts_all%>% unite(\"taxagen\",t3:t8,remove = FALSE) \n",
    "\n",
    "#filter out unwanted functions\n",
    "taxa_wgs_all = taxa_wgs_all %>% filter(!str_detect(X3,\"Octaheme*\"))\n",
    "taxa_wgs_all = taxa_wgs_all %>% filter(!str_detect(X3,\"nitrile*\"))\n",
    "taxa_wgs_all = taxa_wgs_all %>% filter(!str_detect(X3,\"calcium\"))\n",
    "taxa_wgs_all = taxa_wgs_all %>% filter(!str_detect(X3,\"PQQGDH\"))\n",
    "taxa_wgs_all = taxa_wgs_all %>% filter(!str_detect(X3,\"Phosphodiesterase*\"))\n",
    "taxa_wgs_all = taxa_wgs_all %>% filter(!str_detect(X3,\"disproportionation\"))\n",
    "taxa_wgs_all = taxa_wgs_all %>% filter(!str_detect(X3,\"urease\"))\n",
    "\n",
    "taxa_wts_all = taxa_wts_all %>% filter(!str_detect(X3,\"Octaheme*\"))\n",
    "taxa_wts_all = taxa_wts_all %>% filter(!str_detect(X3,\"nitrile*\"))\n",
    "taxa_wts_all = taxa_wts_all %>% filter(!str_detect(X3,\"calcium\"))\n",
    "taxa_wts_all = taxa_wts_all %>% filter(!str_detect(X3,\"PQQGDH\"))\n",
    "taxa_wts_all = taxa_wts_all %>% filter(!str_detect(X3,\"Phosphodiesterase*\"))\n",
    "taxa_wts_all = taxa_wts_all %>% filter(!str_detect(X3,\"disproportionation\"))\n",
    "taxa_wts_all = taxa_wts_all %>% filter(!str_detect(X3,\"urease\"))\n",
    "\n",
    "# sulfur annotation (sequence of rdsr and dsr are highly similar, can be separated into different clusters by aligning to the reference and plot tree)\n",
    "taxa_wgs_all[,13:48] <- lapply(taxa_wgs_all[,13:48], as.numeric)\n",
    "taxa_wts_all[,13:42] <- lapply(taxa_wts_all[,13:42], as.numeric)\n",
    "\n",
    "###get taxonomy node from R data\n",
    "rdsrA<-(taxa_wts_all[grep(\"dsrA\",taxa_wts_all$X4), ])\n",
    "rdsrA<- mutate(rdsrA,total=rowSums(rdsrA[13:42]))\n",
    "rdsrA<-rdsrA[order(rdsrA$total,decreasing = T),]\n",
    "rdsrA1<-as.data.frame(rdsrA[1:1294,1])\n",
    "write_delim(x = rdsrA1,\"rdsrA1.wts.txt\")\n",
    "\n",
    "rdsrA<-(taxa_wgs_all[grep(\"dsrA\",taxa_wgs_all$X4), ])\n",
    "rdsrA<- mutate(rdsrA,total=rowSums(rdsrA[13:48]))\n",
    "rdsrA<-rdsrA[order(rdsrA$total,decreasing = T),]\n",
    "rdsrA1<-as.data.frame(rdsrA[1:2077,1])\n",
    "write_delim(x = rdsrA1,\"rdsrA1.wgs.txt\")\n",
    "\n",
    "rdsrB<-(taxa_wts_all[grep(\"dsrB\",taxa_wts_all$X4), ])\n",
    "rdsrB<- mutate(rdsrB,total=rowSums(rdsrB[13:42]))\n",
    "rdsrB<-rdsrB[order(rdsrB$total,decreasing = T),]\n",
    "rdsrB1<-as.data.frame(rdsrB[1:1351,1])\n",
    "write_delim(x = rdsrB1,\"rdsrB1.wts.txt\")\n",
    "\n",
    "rdsrB<-(taxa_wgs_all[grep(\"dsrB\",taxa_wgs_all$X4), ])\n",
    "rdsrB<- mutate(rdsrB,total=rowSums(rdsrB[13:48]))\n",
    "rdsrB<-rdsrB[order(rdsrB$total,decreasing = T),]\n",
    "rdsrB1<-as.data.frame(rdsrB[1:2176,1])\n",
    "write_delim(x = rdsrB1,\"rdsrB1.wgs.txt\")\n",
    "\n",
    "#then use seqtk in terminal to get sequence and download swissprot sequence as reference\n",
    "cat rdsrA1.w* |sort -u > rdsrA1.txt\n",
    "cat rdsrB1.w* |sort -u > rdsrB1.txt\n",
    "module load seqtk/1.2-gimkl-2017a\n",
    "seqtk subseq all.scaffold.faa rdsrA1.txt >rdsrA.fasta\n",
    "seqtk subseq all.scaffold.faa rdsrB1.txt >rdsrB.fasta\n",
    "\n",
    "#add reference from swissprot and uniprot\n",
    "cat ref*dsrA*.fasta rdsrA.fasta > rdsrA1.fasta\n",
    "cat ref*dsrB*.fasta rdsrB.fasta > rdsrB1.fasta\n",
    "\n",
    "#then align with mafft and get ML tree with fasttree\n",
    "module load MAFFT/7.310-gimkl-2017a\n",
    "mafft --thread 24 --maxiterate 100 rdsrA1.fasta >align.rdsrA.fasta\n",
    "mafft --thread 24 --maxiterate 100 rdsrB1.fasta >align.rdsrB.fasta\n",
    "\n",
    "#edit header as some of them have duplicate node name from different samples\n",
    "sed -i 's/\\(.*_[0-9]*\\) \\# \\([0-9]*\\) \\# \\([0-9]*\\).*/\\1__\\2__\\3/g' align.rdsrA.fasta\n",
    "sed -i 's/\\(.*_[0-9]*\\) \\# \\([0-9]*\\) \\# \\([0-9]*\\).*/\\1__\\2__\\3/g' align.rdsrB.fasta\n",
    "\n",
    "module load FastTree/2.1.10-gimkl-2018b\n",
    "export OMP_NUM_THREADS=12\n",
    "FastTree -gamma align.rdsrA.fasta > align1.rdsrA.tree\n",
    "FastTree -gamma align.rdsrB.fasta > align.rdsrB.tree\n",
    "\n",
    "\n",
    "#input into ITOL (https://itol.embl.de/itol.cgi), look at the tree branch/cluster, and copy node name that are with rdsr cluster\n",
    "#edit the node name to match the real node name\n",
    "sed -i 's/\\(.*cov_[0-9]*.[0-9]*_[0-9]*\\)_\\([0-9]*\\)_\\([0-9]\\)/\\1\\t\\2\\t\\3/g' node.rdsrA.txt\n",
    "sed -i 's/\\(.*cov_[0-9]*.[0-9]*_[0-9]*\\)_\\([0-9]*\\)_\\([0-9]\\)/\\1\\t\\2\\t\\3/g' node.rdsrA.txt\n",
    "\n",
    "#then run in r\n",
    "taxa_wgs_all$X4 = as.character(taxa_wgs_all$X4)\n",
    "taxa_wts_all$X4 = as.character(taxa_wts_all$X4)\n",
    "\n",
    "node_rdsrA<-read.delim(\"node.rdsrA.txt\",header = F)\n",
    "node_rdsrB<-read.delim(\"node.rdsrB.txt\",header = F)\n",
    "\n",
    "\n",
    "taxa_wgs_all = taxa_wgs_all %>% mutate( X4 = case_when( node %in% node_rdsrA$V1 & n7 %in% node_rdsrA$V2~ 'rdsrA', node %in% node_rdsrB$V1 & n7 %in% node_rdsrB$V2~ 'rdsrB', TRUE ~ X4 ) )  \n",
    "taxa_wts_all = taxa_wts_all %>% mutate( X4 = case_when( node %in% node_rdsrA$V1 & n7 %in% node_rdsrA$V2~ 'rdsrA', node %in% node_rdsrB$V1 & n7 %in% node_rdsrB$V2~ 'rdsrB', TRUE ~ X4 ) ) \n",
    "taxa_wgs_all = taxa_wgs_all %>% mutate( X3 = case_when( X4 == \"dsrA\" ~ 'Sulfite_reduction', X4 ==\"dsrB\" ~ 'Sulfite_reduction', TRUE ~ X3 ) )\n",
    "taxa_wts_all = taxa_wts_all %>% mutate( X3 = case_when( X4 == \"dsrA\" ~ 'Sulfite_reduction', X4 ==\"dsrB\" ~ 'Sulfite_reduction', TRUE ~ X3 ) )\n",
    "\n",
    "# Final writing to output\n",
    "taxa_wgs_all[,13:48] <- lapply(taxa_wgs_all[,13:48], as.character)\n",
    "taxa_wts_all[,13:42] <- lapply(taxa_wts_all[,13:42], as.character)\n",
    "\n",
    "taxa_wgs_all[,13:48] <- lapply(taxa_wgs_all[,13:48], as.numeric)\n",
    "taxa_wts_all[,13:42] <- lapply(taxa_wts_all[,13:42], as.numeric)\n",
    "\n",
    "\n",
    "write_delim(taxa_wgs_all,\"taxa_wgs_all.txt\")\n",
    "write_delim(taxa_wts_all,\"taxa_wts_all.txt\")\n",
    "\n",
    "\n",
    "## manual checking for SAR group== diatoms using kraken\n",
    "###get taxonomy node from R data\n",
    "SAR_photo<-with(taxa_wts_all,taxa_wts_all[grepl(\"Sar\",taxa_wts_all$t4) & grepl(\"Photosynthesis\", taxa_wts_all$X2), ])\n",
    "SAR_photo<- mutate(SAR_photo,total=rowSums(SAR_photo[13:42]))\n",
    "SAR_photo<-SAR_photo[order(SAR_photo$total,decreasing = T),]\n",
    "SAR_photo1<-as.data.frame(SAR_photo[1:1302,7])\n",
    "write_delim(x = SAR_photo1,\"sar_photosynthesis.txt\")                                            \n",
    "    \n",
    "SAR_car<-with(taxa_wts_all,taxa_wts_all[grepl(\"Sar\",taxa_wts_all$t4) & grepl(\"Carbon fixation\", taxa_wts_all$X2), ])\n",
    "SAR_car<- mutate(SAR_car,total=rowSums(SAR_car[13:42]))\n",
    "SAR_car<-SAR_car[order(SAR_car$total,decreasing = T),]\n",
    "SAR_car1<-as.data.frame(SAR_car[1:71,7])\n",
    "write_delim(x = SAR_car1,\"sar_carfix.txt\")     \n",
    "                                            \n",
    "module load seqtk/1.2-gimkl-2017a\n",
    "seqtk subseq all.scaffold.fasta sar_photosynthesis.txt >sar_photosynthesis.fasta                        \n",
    "seqtk subseq all.scaffold.fasta sar_carfix.txt >sar_carfix.fasta \n",
    "\n",
    "module load Kraken2/2.0.8-beta-gimkl-2018b\n",
    "kraken2 --db  ~/database/Databases/Kraken2/ --threads 16 sar_carfix.fasta --report taxa.sar.carfix.txt --use-mpa-style --use-name >taxa.sar.carfix.d.txt\n",
    "kraken2 --db  ~/database/Databases/Kraken2/ --threads 16 sar_photosynthesis.fasta --report taxa.sar.photo.txt --use-mpa-style --use-name >taxa.sar.photo.d.txt\n",
    "                                            \n",
    "#count no occurance\n",
    "n_occur <- data.frame(table(taxa_wgs_all$node))\n",
    "n_occur <- data.frame(table(taxa_wts_all$node))\n",
    "\n",
    "\n",
    "#select the value out\n",
    "group_wgs<-taxa_wgs_all %>% select(c(X2,X3,X4,f1:S9R3))\n",
    "group_wts<-taxa_wts_all %>% select(c(X2,X3,X4,f1:S7R3))\n",
    "group_wgs<-group_by(group_wgs,X2,X3,X4)\n",
    "group_wts<-group_by(group_wts,X2,X3,X4)\n",
    "\n",
    "\n",
    "#Change to numeric\n",
    "group_wgs[,4:12]<-lapply(group_wgs[,4:12],as.numeric)\n",
    "group_wts[,4:12]<-lapply(group_wts[,4:12],as.numeric)\n",
    "\n",
    "\n",
    "#rename column names\n",
    "colnames(group_wgs)<-c(\"PATHWAY\",\"FUNCTION\",\"GENE\",\"f1\",\"f2\",\"f3\",\"f4\",\"f5\",\"f6\",\"f7\",\"f8\",\"f9\",\"S1R1\",\"S1R2\",\"S1R3\",\"S2R1\",\"S2R2\",\"S2R3\",\"S3R1\",\"S3R2\",\"S3R3\",\"S4R1\",\"S4R2\",\"S4R3\",\"S5R1\",\"S5R2\",\"S5R3\",\"S6R1\",\"S6R2\",\"S6R3\",\"S7R1\",\"S7R2\",\"S7R3\",\"S8R1\",\"S8R2\",\"S8R3\",\"S9R1\",\"S9R2\",\"S9R3\")\n",
    "colnames(group_wts)<-c(\"PATHWAY\",\"FUNCTION\",\"GENE\",\"f1\",\"f2\",\"f3\",\"f4\",\"f5\",\"f6\",\"f7\",\"f8\",\"f9\",\"S1R1\",\"S1R2\",\"S1R3\",\"S2R1\",\"S2R2\",\"S2R3\",\"S3R1\",\"S3R2\",\"S3R3\",\"S4R1\",\"S4R2\",\"S4R3\",\"S5R1\",\"S5R2\",\"S5R3\",\"S6R1\",\"S6R2\",\"S6R3\",\"S7R1\",\"S7R2\",\"S7R3\")\n",
    "\n",
    "#summarise by each column by group\n",
    "group_wgs<-summarise(group_wgs,sum(f1),sum(f2),sum(f3),sum(f4),sum(f5),sum(f6),sum(f7),sum(f8),sum(f9),sum(S1R1),sum(S1R2),sum(S1R3),sum(S2R1),sum(S2R2),sum(S2R3),sum(S3R1),sum(S3R2),sum(S3R3),sum(S4R1),sum(S4R2),sum(S4R3),sum(S5R1),sum(S5R2),sum(S5R3),sum(S6R1),sum(S6R2),sum(S6R3),sum(S7R1),sum(S7R2),sum(S7R3),sum(S8R1),sum(S8R2),sum(S8R3),sum(S9R1),sum(S9R2),sum(S9R3))\n",
    "group_wts<-summarise(group_wts,sum(f1),sum(f2),sum(f3),sum(f4),sum(f5),sum(f6),sum(f7),sum(f8),sum(f9),sum(S1R1),sum(S1R2),sum(S1R3),sum(S2R1),sum(S2R2),sum(S2R3),sum(S3R1),sum(S3R2),sum(S3R3),sum(S4R1),sum(S4R2),sum(S4R3),sum(S5R1),sum(S5R2),sum(S5R3),sum(S6R1),sum(S6R2),sum(S6R3),sum(S7R1),sum(S7R2),sum(S7R3))\n",
    "\n",
    "\n",
    "write_csv(group_wgs,\"tpm_group_wgs_all.csv\")\n",
    "write_csv(group_wts,\"tpm_group_wts_all.csv\")\n",
    "\n",
    "\n",
    "\n",
    "# For piechart\n",
    "\n",
    "taxa_wgs_all1=cbind((taxa_wgs_all[,1:21]),(t(apply(taxa_wgs_all[,22:48], 1, tapply, gl(9, 3), mean, na.rm = TRUE))),(taxa_wgs_all[,49:76])) %>% as.data.frame()\n",
    "taxa_wts_all1=cbind((taxa_wts_all[,1:21]),t(apply(taxa_wts_all[,22:42], 1, tapply, gl(7, 3), mean, na.rm = TRUE)),(taxa_wts_all[,43:70]))  %>% as.data.frame()\n",
    "\n",
    "group_wgs1<-taxa_wgs_all1 %>% select(c(X2,f1:'9'))\n",
    "group_wts1<-taxa_wts_all1 %>% select(c(X2,f1:'7'))\n",
    "group_wgs1<-group_by(group_wgs1,X2)\n",
    "group_wts1<-group_by(group_wts1,X2)\n",
    "#summarise by each column by group\n",
    "group_wgs1<-summarise(group_wgs1,sum(f1),sum(f2),sum(f3),sum(f4),sum(f5),sum(f6),sum(f7),sum(f8),sum(f9),sum(`1`),sum(`2`),sum(`3`),sum(`4`),sum(`5`),sum(`6`),sum(`7`),sum(`8`),sum(`9`))\n",
    "group_wts1<-summarise(group_wts1,sum(f1),sum(f2),sum(f3),sum(f4),sum(f5),sum(f6),sum(f7),sum(f8),sum(f9),sum(`1`),sum(`2`),sum(`3`),sum(`4`),sum(`5`),sum(`6`),sum(`7`))\n",
    "write_csv(group_wgs1,\"tpm_group_X2wgs_all.csv\")\n",
    "write_csv(group_wts1,\"tpm_group_X2wts_all.csv\")\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
